notifier/notify_windows.go

```diff
@@ -1,16 +1,290 @@
+import (
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"syscall"
 
-func actionListener() {
+	"github.com/Microsoft/go-winio"
+	"github.com/safing/portbase/log"
+	"golang.org/x/text/encoding/unicode"
+)
+
+var (
+	notifierPath = ""
+	logoLocation = ""
+)
 
+const (
+	pipeName        = "\\\\.\\pipe\\portmasterNotifierToast"
+	webLogoLocation = "http://127.0.0.1:817/ui/modules/assets/favicons/ms-icon-310x310.png" //TODO
+)
+
+func init() {
+	if logoLocation == "" {
+		logoLocation = filepath.Join(os.TempDir(), "Portmaster", "logo-310x310.png")
+	}
+}
+
+//API called functions:
+
+func actionListener() {
+	initNotifierPath()
+	go notificationListener()
+	initLogo()
 }
 
 // Show shows the notification.
 func (n *Notification) Show() {
-	log.Warningf("notify: showing notifications not yet implemented. (%s)", n.ID)
+	log.Debugf("called to show: %+v", n)
+	if n.Message == "" {
+		log.Errorf("Snoretoast was asked to create this notification with empty messag - impossible: %+v", n)
+		return
+	}
+	initLogo()
+	snoreToastRunCmdAsync(exec.Command(notifierPath, "-id", n.GUID, "-t", "Portmaster", "-m", n.Message, "-b", buildSnoreToastButtonArgument(n.AvailableActions), "-appID", "io.safing.portmaster", "-p", logoLocation, "-pipeName", pipeName))
 }
 
 // Cancel cancels the notification.
 func (n *Notification) Cancel() {
+	if n == nil || n.GUID == "" {
+		return
+	}
+	snoreToastRunCmdAsync(exec.Command(notifierPath, "-close", n.GUID, "-appID", "io.safing.portmaster"))
+}
+
+//internal functions:
+
+func initNotifierPath() {
+	if notifierPath == "" {
+		var err error
+		notifierPath, err = getPath("notifier-snoretoast")
+
+		if err != nil {
+			log.Errorf("Error while getting SnoreToast-Path: %s %s", err, notifierPath)
+			return
+			//TODO: what to do?
+		}
+	}
+}
+
+func initLogo() {
+	if _, err := os.Stat(logoLocation); err != nil { //File doesn't exist or another Error that is handled while copying file there
+		os.MkdirAll(filepath.Dir(logoLocation), os.ModePerm)
+		if err = downloadFile(logoLocation, webLogoLocation); err != nil {
+			log.Errorf("Can't copy Logo for SnoreToast from %s to %s: %s", webLogoLocation, logoLocation, err)
+		}
+	}
+}
+
+func notificationListener() {
+	log.Debugf("Listening on Pipe %s\n", pipeName)
+	l, err := winio.ListenPipe(pipeName, nil)
+	if err != nil {
+		log.Errorf("Error while starting Pipe for SnoreToast: %s", err)
+		return
+	}
+	defer l.Close()
+
+	for {
+		conn, err := l.Accept()
+		if err != nil {
+			log.Errorf(err)
+			continue
+		}
+		go onCallback(conn)
+	}
+}
+
+func onCallback(conn net.Conn) {
+	data := snoreToastStrToMap(readWstrConnToStr(conn)) //TODO: what  if some required attribute isn't there?
+	log.Debugf("onCallback: %+v", data)
+
+	if data["action"] == "timedout" {
+		return
+	}
+
+	notificationsLock.Lock()
+	var n *Notification
+	for _, elem := range notifications {
+		if elem.GUID == data["notificationId"] {
+			n = elem
+		}
```

Comment:
Hint: break the loop as soon as you have found the desired entry of the array/slice.